pkg/decoder/decoder.go

```diff
@@ -7,102 +7,205 @@ package decoder
+		msgBuffer:  &msgBuffer,
+		lineBuffer: &lineBuffer,
+		lineRe:     lineRe,
 	}
 }
 
+// isMultiLineEnabled returns true if the decoder supports multi-line logs
+func (d *Decoder) isMultiLineEnabled() bool {
+	return d.lineRe != nil
+}
+
 // Start starts the Decoder
 func (d *Decoder) Start() {
 	go d.run()
 }
 
+// Stop stops the Decoder
+func (d *Decoder) Stop() {
+	close(d.InputChan)
+}
+
 // run lets the Decoder handle data coming from the InputChan
 func (d *Decoder) run() {
 	for data := range d.InputChan {
-		d.decodeIncomingData(data.content, data.offset)
+		if d.isMultiLineEnabled() {
+			d.ackIncomingData()
+			endedWithEOL := d.decodeIncomingData(data.content, data.context)
+			if endedWithEOL {
+				d.ackEndIncomingData(data.context)
+			}
+		} else {
+			d.decodeIncomingData(data.content, data.context)
+		}
+
 	}
 	d.OutputChan <- message.NewStopMessage()
 }
 
-// Stop stops the Decoder
-func (d *Decoder) Stop() {
-	close(d.InputChan)
+// ackIncomingData stops the timer which flushes the multi-line buffer as we got new data
+func (d *Decoder) ackIncomingData() {
+	if d.timer != nil {
+		d.timer.Stop()
+	}
+}
+
+// ackEndIncomingData starts the timer which flushes the multi-line buffer
+func (d *Decoder) ackEndIncomingData(context PayloadContext) {
+	d.timer = time.AfterFunc(countDownDuration, func() {
+		d.sendMessage(context)
+	})
 }
 
-var truncatedMsg = []byte("...TRUNCATED...")
-var truncatedLen = len(truncatedMsg)
-var maxMessageLen = config.MaxMessageLen - truncatedLen
+// decodeIncomingData splits raw data based on `\n`, creates and processes new lines
+// returns true if inBuf ends with `\n`
+func (d *Decoder) decodeIncomingData(inBuf []byte, context PayloadContext) (endsWithEOL bool) {
+	i, j := 0, 0
+	n := len(inBuf)
+	maxj := i + maxMessageLen - d.lineBuffer.Len()
 
-// sendBuffuredMessage flushes the buffer and sends the message
-func (d *Decoder) sendBuffuredMessage(offset int64) {
+	for ; j < n; j++ {
+		if j == maxj {
+			// process the line as it is too long
+			d.lineBuffer.Write(inBuf[i:j])
+			d.processLine(context)
+			i = j
+			maxj = i + maxMessageLen
+		} else if inBuf[j] == '\n' {
+			d.lineBuffer.Write(inBuf[i:j])
+			d.processLine(context)
+			i = j + 1 // +1 as we skip the `\n`
+			maxj = i + maxMessageLen
+		}
+	}
+	d.lineBuffer.Write(inBuf[i:j])
+
+	// check if inBuf ends with `\n`
+	if len(inBuf) > 0 && inBuf[n-1] == '\n' {
+		endsWithEOL = true
+	}
+	return endsWithEOL
+}
+
+// processLine appends new line to the message, sends and truncates messages
+func (d *Decoder) processLine(context PayloadContext) {
+	line := d.lineBuffer.Bytes()
+	defer d.lineBuffer.Reset()
+
+	isLineAdded := false
+	if d.isMultiLineEnabled() {
+		if d.lineRe.Match(line) {
+			d.sendMessage(context)
+		}
+		isLineAdded = d.appendLine(line, context)
+	} else {
+		isLineAdded = d.appendLine(line, context)
+		if isLineAdded {
+			d.sendMessage(context)
+		}
+	}
+	if !isLineAdded {
+		d.truncateAndSendMessage(line, context)
+	}
```

Comment:
I find `isLineAdded` slightly confusing, what about using an error for that, and checking on error being `buffer full` for instance?